frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -2448,19 +2472,51 @@ const IR::Node* TypeInference::postorder(IR::Member* expression) {
                 return expression;
             }
         }
-        if (type->is<IR::Type_Header>()) {
-            if (member == IR::Type_Header::setValid ||
-                member == IR::Type_Header::setInvalid) {
+        if (type->is<IR::Type_Header>() || type->is<IR::Type_HeaderUnion>()) {
+            if (type->is<IR::Type_HeaderUnion>() &&
```

Comment:
This code is not necessary. You get a fine error message without it.

Reply:
Another reviewer asked to add the new error.  It makes sense because now HeaderUnion is also allowed (line 2473) and one should check for setValid() and setInvalid() not supported by HeaderUnion